File: exam_in_class/2/c_is_for_cycle.cc
/*
* author: redleaf
* email: redleaf23477@gapp.nthu.edu.tw
*/
#include <bits/stdc++.h>
using namespace std;
using LL = long long int;

int main() {
    ios::sync_with_stdio(false); cin.tie();
    int n, m; cin >> n >> m;
    vector<vector<int>> graph(n);
    for (int i = 0; i < m; i++) {
        int u, v; cin >> u >> v; u--, v--;
        graph[u].emplace_back(v);
        graph[v].emplace_back(u);
    }
    vector<bool> vis(n, false);
    vector<int> stk;
    function<bool(int,int)> dfs = [&](int u, int pa) {
        vis[u] = true;
        stk.emplace_back(u);
        for (auto v : graph[u]) {
            if (v == pa) continue;
            if (vis[v]) {
                auto i = find(stk.begin(), stk.end(), v) - stk.begin();
                cout << stk.size() - i << "\n";
                for (; i < stk.size(); i++) {
                    cout << stk[i] + 1 << " ";
                }
                cout << "\n";
                return true;
            } else {
                if (dfs(v, u)) return true;
            }
        }
        stk.pop_back();
        return false;
    };
    bool ans = false;
    for (int i = 0; i < n && !ans; i++) {
        if (!vis[i]) {
            ans = dfs(i, i);
        }
    }
    if (!ans) {
        cout << "IMPOSSIBLE\n";
    }
}

/*
* author: jamie
* email: jinjiedeng.jjd@gmail.com
*/
#include <iostream>
#include <vector>

bool IsCycleExisted
(
  std::vector<std::vector<uint64_t>> const& graph,
  std::vector<bool>& is_visited,
  uint64_t parent_from,
  uint64_t from,
  std::vector<uint64_t>& cycle
)
{
  static uint64_t detected_node {};
  is_visited[from] = true;
  for (auto to : graph[from])
  {
    if (!is_visited[to])
    {
      if (IsCycleExisted(graph, is_visited, from, to, cycle))
      {
        if (detected_node)
        {
          if (from == detected_node) { detected_node = 0; }
          cycle.push_back(from);
        }
        return true;
      }
    }
    else if (to != parent_from)
    {
      detected_node = to;
      cycle.push_back(from);
      return true;
    }
  }
  return false;
}

int main ()
{
  uint64_t n, m;
  std::cin >> n >> m;
  std::vector<std::vector<uint64_t>> graph(n + 1);
  while (m--)
  {
    uint64_t from, to;
    std::cin >> from >> to;
    graph[from].emplace_back(to);
    graph[to].emplace_back(from);
  }
  std::vector<bool> is_visited(n + 1);
  std::vector<uint64_t> cycle;
  for (uint64_t source {1}; source != graph.size(); ++source)
  {
    if (!is_visited[source])
    {
      if (IsCycleExisted(graph, is_visited, 0, source, cycle)) { break; }
    }
  }
  if (!cycle.empty())
  {
    std::cout << cycle.size() << "\n";
    for (auto v : cycle) { std::cout << v << " "; }
  }
  else { std::cout << "IMPOSSIBLE"; }
  return 0;
}

File: exam_in_class/2/c_is_for_cycle.py
# author: redleaf
# email: redleaf23477@gapp.nthu.edu.tw
def dfs(s, n, graph, vis):
    '''
    iterative version of dfs
    '''
    stack = [s]
    vis[s] = True
    while len(stack) > 0:
        u = stack[-1]

        # pop parent node in dfs tree
        if len(graph[u]) > 0 and len(stack) > 1 and graph[u][-1] == stack[-2]:
            graph[u].pop()

        # if all neighbors of u doesn't lead to a cycle, bye u
        if len(graph[u]) == 0:
            stack.pop()
            continue

        # get an neighbor
        v = graph[u].pop()

        if vis[v] == True:
            # I found a cycle v -> ... -> u -> v
            idx = stack.index(v)
            cycle = stack[idx:]
            return cycle
        else:
            # add tree edge in DFS tree
            vis[v] = True
            stack.append(v)

    # no cycle found
    return []


def find_cycle(n, graph):
    vis = [False] * n
    for u in range(n):
        if vis[u] == False:
            c = dfs(u, n, graph, vis)
            if len(c) > 0:
                return c
    return []


if __name__ == '__main__':
    n, m = map(int, input().split())
    graph = [[] for _ in range(n)]
    for _ in range(m):
        u, v = map(int, input().split())
        u, v = u - 1, v - 1
        graph[u].append(v)
        graph[v].append(u)

    cycle = find_cycle(n, graph)
    if len(cycle) == 0:
        print("IMPOSSIBLE")
    else:
        print(len(cycle))
        print(*[x + 1 for x in cycle])

# author: jamie
# email: jinjiedeng.jjd@gmail.com
def IsCycleExisted (graph, is_visited, parent_u, u, cycle):
    global detected_node
    is_visited[u] = True
    for v in graph[u]:
        if not is_visited[v]:
            is_visited[v] = True
            if IsCycleExisted(graph, is_visited, u, v, cycle):
                if detected_node:
                    cycle.append(u)
                    if detected_node == u:
                        detected_node = 0
                return True
        elif v != parent_u:
            cycle.append(u)
            detected_node = v
            return True
    return False

if __name__ == "__main__":
    n, m = map(int, input().split())
    graph = [[] for _ in range(n + 1)]
    for _ in range(m):
        u, v = map(int, input().split())
        graph[u].append(v)
        graph[v].append(u)
    is_visited = [False] * (n + 1)
    cycle = []
    for source in range(1, n + 1):
        if not is_visited[source]:
            detected_node = 0
            if IsCycleExisted(graph, is_visited, 0, source, cycle):
                break
    if len(cycle):
        print(len(cycle))
        print(" ".join(map(str, cycle)))
    else:
        print("IMPOSSIBLE")